### PR TITLE
IOSXE/show cts ap sgt info {ap_name}

### PR DESCRIPTION
**Pull request by qduk**
## Description
Created parser for IOSXE/show cts ap sgt info {ap_name}

## Motivation and Context
Able to parse unstructured output from show cts ap sgt info {ap_name} to structured data

## Screenshots:
Capture

## Checklist:
- [x] I have updated the changelog.
- [x] All new and existing tests passed.
- [x] All new code passed compilation.

### CODE REVIEW

File: src/genie/libs/parser/iosxe/show_cts.py [hunk 1]
```diff
@@ -1945,3 +1945,79 @@ def filter_lines(raw_output):
...
+                continue
+            elif p_sgt_header.match(line):
+                continue
+            elif p_sgt_delimiter.match(line):
+                continue
+            elif p_sgt_row.match(line):
+                match = p_sgt_row.match(line)
+                group = match.groupdict()
+                if not cts_ap_dict.get("sgts"):
+                    cts_ap_dict.update({ "sgts": {} })
```
Comment (danielgraziano) [inline src/genie/libs/parser/iosxe/show_cts.py:2018]: No big deal if not changed, however you can do this instead
``` python
sgts_dict = cts_ap_dict.setdefault("sgts", {})  # only sets the key/value if the key doesnt already exist
sgts_dict.update(
	{
		group["sgt"]: {
			"policy_pushed_to_ap": group["policy"].lower(), 
			"no_of_clients": int(group["clients"])
		}
	}
)

**Review by danielgraziano — COMMENTED** [review]
Overall looks good. Lets wait for schema review from @tahigash

File: src/genie/libs/parser/iosxe/show_cts.py [hunk 2]
```diff
@@ -1945,3 +1945,79 @@ def filter_lines(raw_output):
...
+class ShowCtsApSgtInfo(ShowCtsApSgtInfoSchema):
+    """Parser for show cts ap sgt info {ap_name}"""
+
+    cli_command = 'show cts ap sgt info {ap_name}'
+
+    def cli(self, ap_name="", output=None):
+        if output is None:
+            output = self.device.execute(self.cli_command.format(ap_name=ap_name))
+        else:
+            output = output
```
[the remark this answers is not in the document]
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_cts.py]: Updated

File: src/genie/libs/parser/iosxe/show_cts.py [hunk 3]
```diff
@@ -1945,3 +1945,79 @@ def filter_lines(raw_output):
...
+
+        # Number of SGTs referred by the AP...............: 2
+        #
+        # SGT               PolicyPushedToAP       No.of Clients
+        # ------------------------------------------------------------
+        # UNKNOWN(0)        NO                     0
+        # DEFAULT(65535)    NO                     0
+
+        # Number of SGTs referred by the AP...............: 0
+        p_number_sgts = re.compile(r"^Number\s+of\s+SGTs\s+referred\s+by\s+the\s+AP...............:\s+(?P<value>\d+)$")
```
[the remark this answers is not in the document]
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_cts.py]: Good call. Updated!

File: src/genie/libs/parser/iosxe/show_cts.py [hunk 4]
```diff
@@ -1945,3 +1945,79 @@ def filter_lines(raw_output):
...
+
+        # SGT               PolicyPushedToAP       No.of Clients
+        p_sgt_header = re.compile(r"^SGT\s+PolicyPushedToAP\s+No.of\s+Clients$")
+
+        # ------------------------------------------------------------
+        p_sgt_delimiter = re.compile(r"^------------------------------------------------------------$")
```
[the remark this answers is not in the document]
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_cts.py]: Updated

File: src/genie/libs/parser/iosxe/show_cts.py [hunk 5]
```diff
@@ -1945,3 +1945,79 @@ def filter_lines(raw_output):
...
+
+        # 10                NO                     1
+        p_sgt_row = re.compile(r"^(?P<sgt>\S+)\s+(?P<policy>\S+)\s+(?P<clients>\d+)$")
+
+        cts_ap_dict = {}
+
+        for line in output.splitlines():
+            line = line.strip()
+            if p_number_sgts.match(line):
```
[the remark this answers is not in the document]
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_cts.py]: I would've never known it runs faster that one. Thank you! Updated!

File: src/genie/libs/parser/iosxe/show_cts.py (hunk 1 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_cts.py]: I need to get in the habit of using `setdefault()` it is much cleaner. Thank you! Updated!

File: src/genie/libs/parser/iosxe/tests/ShowCtsApSgtInfo/cli/equal/golden_output_expected.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+expected_output = {
+    "number_of_sgts_referred_by_the_ap": 2,
+    "sgts": {
+        "UNKNOWN(0)": {
+            "policy_pushed_to_ap": "no",
```
Comment (tahigash) [inline src/genie/libs/parser/iosxe/tests/ShowCtsApSgtInfo/cli/equal/golden_output_expected.py:5]: please take the value as is.
```suggestion
            "policy_pushed_to_ap": "NO",
```

File: src/genie/libs/parser/iosxe/show_cts.py [hunk 7]
```diff
@@ -1945,3 +1945,79 @@ def filter_lines(raw_output):
...
+
+
+# =========================
+# Schema for:
+#  * 'show cts ap sgt info {ap_name}'
+# =========================
+class ShowCtsApSgtInfoSchema(MetaParser):
+    """Schema for show cts ap sgt info {ap_name}."""
+
+    schema = {
```
[the remark this answers is not in the document]
Comment (tahigash) [reply src/genie/libs/parser/iosxe/show_cts.py:1957]: Please update schema as per @danielgraziano suggestion.

**Review by tahigash — CHANGES_REQUESTED** [review]
reviewed schema and left comments

File: src/genie/libs/parser/iosxe/tests/ShowCtsApSgtInfo/cli/equal/golden_output_expected.py (hunk 6 above)
Comment (qduk) [reply src/genie/libs/parser/iosxe/tests/ShowCtsApSgtInfo/cli/equal/golden_output_expected.py]: Updated

File: src/genie/libs/parser/iosxe/show_cts.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (qduk) [reply src/genie/libs/parser/iosxe/show_cts.py:1957]: I have updated the schema to reflect the above suggestion. Thanks!

**Review by tahigash — APPROVED** [review]
schema looks good now

**Review by danielgraziano — APPROVED** [review]

**Pull request merged**